### DISCUSSION TITLE
Heat: Pre-Build Harvest from Relative Path

### DISCUSSION THREAD

**Discussion by KMarois** [wixtoolset/issues#7498] [Questions]
I have a VS2022 solution using Wix 3 and a WPF project. I'm trying to harvest from the bin/release folder of the WPF project into the Wix project.

In the Wix project's Pre-Build event I have
`heat dir "C:\Projects\J2\Line2-Win-Next\DesktopPhone\bin\Release ER" -out "C:\Projects\J2\Line2-Win-Next\SetupER\AppFiles.wxs"`

The solution is located at 
`C:\Projects\J2\Line2-Win-Next\`

I want to replace the hard coded paths with relative paths, but this doesn't seem to work. In some cases, it crashes VS
`heat dir "$(var.SolutionDir)\DesktopPhone\bin\Release ER\" -out "$(var.SolutionDir)\SetupER\AppFiles.wxs"`

The hard-coded paths work fine in development, but this needs to run as well via MSBuild on the server.

What's the right way to do this?

**Comment by KMarois** [comment]
I have resolved this by using the $(SolutionDir) variable
`<PreBuildEvent>heat dir "$(SolutionDir)\DesktopPhone\bin\Release ER" -dr INSTALLDIR -cg cpg_AppFiles -gg -sf -srd -sreg -scom -var "var.Line2.TargetDir" -out "$(SolutionDir)\SetupER\AppFiles.wxs"</PreBuildEvent>
`

**Comment by BMurri** [comment]
If you are using `MSBuild`/`Visual Studio` WiX projects, why aren't you using `HarvestDirectory`?

**Reply by MaroisConsulting** [reply]
When I use HarvestDirectory in my VS project, VS crashes during build

I'm open to suggestions

**Reply by BMurri** [reply]
I have not seen that at all. Do a binlog build, get a stack trace, poke around to see the related property and item values (including relevant metadata), and if you still can't figure it out, share here and we'll try to help.

**Reply by MaroisConsulting** [reply]
Will do.

But isn't HarvesDirectoryn doing the same thing as my Pre-Build event?

**Reply by BMurri** [reply]
It does more: its defaults produce better quality WiX code, it performs better argument validation, it adds the outputs into the `Compile` item for you, it can be smart enough to only run if there are changes, and it's sequenced to run during msbuild's resolution steps, meaning that it can take advantage of connections around or outside of the solution in a VS native way.

**Reply by MaroisConsulting** [reply]
So here's the ultimate goal.  We have a github build going, and it's trying to take the output from the solution build and copy it to AWS S3. The Pre-Build approach works fine on my PC, of course, but when I try it on the build machine I'm getting pathing errors.

Can you point me to some documentation for HarvestDirectory? I saw this
https://wixtoolset.org/docs/v3/msbuild/task_reference/heatdirectory/
but is there anything else?

**Reply by BMurri** [reply]
https://wixtoolset.org/docs/v3/msbuild/target_reference/harvestdirectory/

**Reply by MaroisConsulting** [reply]
Where does this go?

`<ItemGroup>
  <HarvestDirectory Include="..\TestProject\Data">
    <DirectoryRefId>DataDir</DirectoryRefId>
  </HarvestDirectory>
</ItemGroup>`

**Reply by BMurri** [reply]
Anywhere directly inside the root `Project` element of the WiX project file.

**Reply by KMarois** [reply]
OK, so do I need Heat installed on the build server?

**Reply by KMarois** [reply]
Well, I think this documentation is really poor.
https://wixtoolset.org/docs/v3/msbuild/target_reference/harvestdirectory/
For DirectoryRefId it says "The ID of the directory to reference instead of TARGETDIR."..  What in the world does that mean???

I need better examples than this. I'll have to go Google all of this

**Reply by robmen** [reply]
You can pick a different `Directory` from `TARGETDIR` by referencing its `Id`, the same as you can when using the `DirectoryRef` element in .wxs code.

What did the text lead you to believe it meant?

Perhaps you could help improve the documentation since it confused you so much.

It makes perfect sense to me... but I'm (obviously) not a beginner using the Windows Installer.

Note: `HarvestDirectory` in WiX v3 is broken in VS2022. It is fixed in WiX v4.

**Reply by KMarois** [reply]
Well, I'm really confused now.

First, I removed the Pre-Build code I had in there.

Next I tried adding HarvestDirectory item group
`<ItemGroup>
  <HarvestDirectory Include="..\MyProject\Release">
    <DirectoryRefId>var.WixDemo.TargetDir</DirectoryRefId>
  </HarvestDirectory>
</ItemGroup>`
I also added additional relevant tags.

This had no effect. No output setup was created.

Third, I removed the HarvestDirectory, so now there's neither the HarvestDirectory  NOR the Pre-Build.... and a Setup.msi is created!!

I don't understand this. If there's no HarvestDirectory and no pre-build, then how does Wix get the files for the setup?

I have this in Product.wsx
`<Feature Id="ProductFeature" Level="1">
    <ComponentGroupRef Id="cpg_AppFiles" />
</Feature>
`

and I have a file called 'AppFiles.wsx'

`<ComponentGroup Id="cpg_AppFiles">
    <Component Id="cmp6ED75BAE4B394A9E5AFD82AEE3123155" Directory="INSTALLDIR" Guid="{F063E11F-02CE-49C0-B251-1C1327311CB1}">
    </Component>`

Does it find the component 'cpg_AppFiles' and automatically populate it with the files?

**Reply by BMurri** [reply]
If you confirmed that the MSI file actually contains the files from the desired folder, I'm going to guess that `$(SolutionDir)\SetupER\AppFiles.wxs` is also in your WiX project file. Since that file exists, it doesn't have to be recreated every single build (which is the original design impetus for creating the `heat` tool in the first place). Otherwise, the component will be in the MSI but no files will be assigned to it.

**Reply by KMarois** [reply]
> Note: `HarvestDirectory` in WiX v3 is broken in VS2022. It is fixed in WiX v4.
I'm using VS22 .So how do I harvest the files from my source project?

**Comment by BryanCrotaz** [comment]
@KMarois did you solve this? I'm going round in similar circles to you , except that I haven't managed to have the same luck as your "and a Setup.msi is created!!" - my msi is only 32k so I guess it's empty.

**Reply by BryanCrotaz** [reply]
> For DirectoryRefId it says "The ID of the directory to reference instead of TARGETDIR.".. What in the world does that mean???

I had exactly the same thought. Does it mean the directory to read the files from, or the directory to write them to? Where does the variable TARGETDIR (I assume this is the default?) get set? 

Another great example is `HarvestDirectoryDirectoryRefId | Optional string property. The identifier of the Directory element that will contain all generated authoring.`

There's a name overlap here (ends in DirectoryRefId), so is this actually the same property? In which case why the two different descriptions? Is this a global version of DirectoryRefId? Seems unlikely, or the local one would have talked about this instead of `TARGETDIR`. And what is authoring? It's mentioned on this page, (I can't find it anywhere else) without saying what it refers to. If this is the global version of DirectoryRefId, the authoring must mean the input files from the source project, which would seem odd (authoring implies it's an output not an input), and if authoring is an output, then I don't understand why the name duplication with an input directory setting.

I'd love to contribute to the docs but I don't think I understand even 10% of this stuff yet. All I wanted to do two weeks ago was build an MSI for an exe and a few dlls... Has anyone in this thread worked this stuff out yet?

**Reply by MaroisConsulting** [reply]
@BryanCrotaz  I created a simple test app where @robmen helped me get it working. [It's here](https://github.com/MaroisConsulting/Wix4SampleApp).

**Reply by BryanCrotaz** [reply]
@MaroisConsulting  That's the same as I had, but I found (after an allnighter and waaaay too much caffeine) that I had to include all the suppression, otherwise I either got loads of errors or nothing at all.
```
<HarvestDirectory Include="$(SolutionDir)AgentService\bin\Release\net7.0">
	<ComponentGroupName>AgentServiceFiles</ComponentGroupName>
	<DirectoryRefId>INSTALLFOLDER</DirectoryRefId>
	<SuppressRootDirectory>true</SuppressRootDirectory>
	<!-- dont know why but disabling harvesting registry keys makes HEAT5151 error go away here -->
	<SuppressRegistry>true</SuppressRegistry>
	<SuppressFragments>true</SuppressFragments>
	<SuppressCom>true</SuppressCom>
</HarvestDirectory>
```

However even though this works on my dev machine, this fails on the build server saying that my component group doesn't exist. And I can't see how to attach ServiceInstall to the component group that HarvetDirectory puts out.

Are you at the stage of understanding this stuff, or is it still magic incantations?

On a sidenote, do you know where Property.ProductName comes from? It's not set in any of your files as far as I can see. I've tried setting it with `SetProperty` but then I get an error with Duplicated ProductName.

**Reply by MaroisConsulting** [reply]
Sorry, but I'm pretty new with Wix also. Your best bet would be to reach out @robmen

**Reply by BMurri** [reply]
Watch the start of @robmen's tutorials on [harvesting in WiX v4](https://www.youtube.com/watch?v=XcUmwegYClg) (then stick around for the rest. The shows are done live weekly).

**Reply by BMurri** [reply]
> However even though this works on my dev machine, this fails on the build server saying that my component group doesn't exist. And I can't see how to attach ServiceInstall to the component group that HarvetDirectory puts out.

That involves XSLT (aka XSL, aka Extensible Stylesheet Language [Transformations]), invented to transform XML into any other structured text, including XML 😉). Like `regex`, it's powerful, difficult to learn, and unforgiving of rookie mistakes. It never gained much popularity, especially in the use case its inventors envisioned (which was NOT the largest use case it has been employed in, which is transforming one XML into another), and the newer versions of the standard (which are MUCH BETTER) were never implemented by any "mainstream" XML library (including Microsoft).

It's easier to remove the binary that needs `ServiceInstall` from harvesting and author it directly, frankly.

**Reply by BryanCrotaz** [reply]
How do you insert an xslt transform in the Wix file, and where is the documentation to say what you need to do with it?
(I've published 500 page PDFs with XSL-FO so I know what I'm doing)

In the alternative, I can't find any mention in the docs of excluding files from harvesting. How do you do it? Eg harvest a directory except for the .eve

**Reply by BMurri** [reply]
> How do you insert an xslt transform in the Wix file, and where is the documentation to say what you need to do with it? (I've published 500 page PDFs with XSL-FO so I know what I'm doing)

[https://wixtoolset.org/docs/tools/heat/](https://wixtoolset.org/docs/tools/heat/)

Under any of the three `Harvest*` elements add a `Transforms` property with the path(s) (relative to the project file) of your `xsl[t]` file(s). That property takes a list of files (usually it's just one) and runs them serially (IIRC, that code was written over a decade ago).

> In the alternative, I can't find any mention in the docs of excluding files from harvesting. How do you do it? Eg harvest a directory except for the .eve

Use your XSLT to not include the files you wish to exclude.

**Reply by BryanCrotaz** [reply]
How do you control the output filename to be able to debug the xslt?

**Reply by MaroisConsulting** [reply]
It's in the installer project's app properties. Right click on the project and go to Properties

**Reply by BryanCrotaz** [reply]
I meant the output from heat

**Reply by BMurri** [reply]
Heat runs the `XSLT` in-memory. No input or output file is directly involved at that stage of the pipeline.

For debugging purposes, the `WXS` files are created under obj and are named based on the "source arguments" of `heat`. Find the one you are interested in, debug your `XSLT` on it, then save your file and give to `heat` to use.

**Reply by BryanCrotaz** [reply]
> For debugging purposes, the WXS files are created under obj 

The files post-xslt you mean?

**Reply by BMurri** [reply]
For every file that `heat` generates, it does all of the things it does, then it runs the resulting `XML` through the transforms provided, then it writes the results to disk. Before that last point, all of the output is in memory.

The XSL processor is the one (most likely) in .Net Framework 4.x, so it has some limitations and quirks. The source code will show the settings applied (I believe it's the only instance of XSL used in the codebase).

**Reply by BryanCrotaz** [reply]
>Right click on the project and go to Properties
...
> then it writes the results to disk.

Which property do I need to set to control the output filename of heat during a HarvestDirectory?

**Reply by BMurri** [reply]
If it's from msbuild/visual studio project file, you don't, because it doesn't matter, it's added to the `Compile` list for you.

If you're calling heat from the command-line, it's the `-o` command-line argument.

**Comment by BryanCrotaz** [comment]
I’m using HarvestDirectory which uses heat under the hood as far as I know.

Obviously to debug I need to know where to find the output wxs of the
HarvestDirectory pre and post xslt. That’s what I’m looking to control so I
can see the output is correct.

Bryan Crotaz
Silver Curve

On 1 Nov 2023, at 21:39, Blair L Murri ***@***.***> wrote:

﻿

If it's from msbuild/visual studio project file, you don't, because it
doesn't matter, it's added to the Compile list for you.

If you're calling heat from the command-line, it's the -o command-line
argument.

—
Reply to this email directly, view it on GitHub
<https://github.com/orgs/wixtoolset/discussions/7498#discussioncomment-7450131>,
or unsubscribe
<https://github.com/notifications/unsubscribe-auth/AAPMKC767SU377OO6Q2MSHDYCK6SLAVCNFSM6AAAAAAYH5GEZCVHI2DSMVQWIX3LMV43SRDJONRXK43TNFXW4Q3PNVWWK3TUHM3TINJQGEZTC>
.
You are receiving this because you were mentioned.Message ID:
***@***.***>

**Reply by BMurri** [reply]
From the same inputs and arguments, the same outputs are produced pre-xslt.

So build without a transform in place, look under the `obj` folder for `wxs` files with names that include parts of the name of the directory, make a copy of that file and use it to craft your transform.

Add your transform in the project file and build again. Compare the new version of that generated file to the copy you made to see what your transform did.

Rinse and repeat.

When provided a transform, heat does not write the input to the transform to disk.

The output is correct if that output meets the needs of your project.

Watch that episode of the Development Dojo that @robmen produced that I previously linked. It shows the process and he makes all files publicly available on GitHub.

**Comment by BryanCrotaz** [comment]
TBH about half an hour ago I finally got my installer working. I gave up on Heat some days ago as it was just too hard to understand with the existing docs, and just hand typed hundreds of lines of config. 

Please form the above discussion into some docs for the next newbie who tries to find their way through the WiX maze so they don't have to do it blindfolded.

**Reply by BMurri** [reply]
The original use case that led to the v3 harvesting tool named `heat` was a "run once, manually maintain forever" model that generally worked well in a framework world where dependencies of dependencies had to be included by the end-developer and so tended to be rarely changed (which is no longer the world we live in today). Experience also showed that use case was not the most desired use case.

The WiX core team has a replacement tool in the works for v5 written for this new use case and for the newer SDK-style toolsets (which the v3 tooling was NOT written for) so those of us who already have this working don't really have enough volunteer hours left to augment the documentation for an implementation of a feature that will go away in a handful of months.

However this discussion will live here as documentation and is searchable, so the next newbie should have a better leg up. I'm glad you have your installer working.

**Reply by BryanCrotaz** [reply]
Can't it just be copy-pasted into the HarvestDirectory doc? That's a job of a few minutes.

**Reply by BMurri** [reply]
Please do. Contributions are always welcome. https://github.com/wixtoolset/web/blob/master/src/Docusaurus/docs/tools/heat.md
